examples/invoker/client.go

```diff
@@ -88,66 +94,68 @@ func main() {
 		defer shutdown()
 	}
 
-	realRPS := runBenchmark(urls, *runDuration, *rps)
+	realRPS := runBenchmark(endpoints, *runDuration, *rps)
 
 	writeLatencies(realRPS, *latencyOutputFile)
 }
 
-func readLines(path string) ([]string, error) {
-	file, err := os.Open(path)
+func readEndpoints(path string) (endpoints []Endpoint, _ error) {
+	data, err := ioutil.ReadFile(path)
 	if err != nil {
 		return nil, err
 	}
-	defer file.Close()
-
-	var lines []string
-	scanner := bufio.NewScanner(file)
-	for scanner.Scan() {
-		lines = append(lines, scanner.Text())
+	if err := json.Unmarshal(data, &endpoints); err != nil {
+		return nil, err
 	}
-	return lines, scanner.Err()
+	return
 }
 
-func runBenchmark(urls []string, runDuration, targetRPS int) (realRPS float64) {
+func runBenchmark(endpoints []Endpoint, runDuration, targetRPS int) (realRPS float64) {
 	timeout := time.After(time.Duration(runDuration) * time.Second)
 	tick := time.Tick(time.Duration(1000/targetRPS) * time.Millisecond)
 
 	var issued int
 	start := time.Now()
 
+	Start(TimeseriesDBAddr, endpoints)
+
+	loop:
 	for {
+		endpoint := endpoints[issued%len(endpoints)]
```

Comment:
I think that invocations still happen upon ticks but this code is very unintuitive. Please go back to the old code structure with the code under the corresponding <-timeout and <-tick

Reply:
I think it's a bad idea as we'll re-introduce a bug: in the previous style, we wait for the first tick to start and when we have started timing much earlier. Therefore our real RPS was **never** equal to 1.0, suggesting as if there was something wrong with the experiment or the infrastructure whereas it was an invoker bug.

You can check [StackOverflow](https://stackoverflow.com/a/54752803/4466589) too---disclaimer: I answered that question. :)